File: asherah-java/java/src/main/java/com/godaddy/asherah/appencryption/SessionFactory.java
package com.godaddy.asherah.appencryption;

import com.godaddy.asherah.appencryption.kms.AwsKeyManagementServiceImpl;
import com.godaddy.asherah.appencryption.kms.KeyManagementService;
import com.godaddy.asherah.appencryption.kms.StaticKeyManagementServiceImpl;
import com.godaddy.asherah.appencryption.persistence.DynamoDbMetastoreImpl;
import com.godaddy.asherah.appencryption.persistence.InMemoryMetastoreImpl;
import com.godaddy.asherah.appencryption.persistence.JdbcMetastoreImpl;
import com.godaddy.asherah.appencryption.persistence.Metastore;
import com.godaddy.asherah.crypto.CryptoPolicy;
import com.godaddy.asherah.jni.Asherah;
import com.godaddy.asherah.jni.AsherahConfig;
import com.godaddy.asherah.jni.AsherahFactory;
import org.json.JSONObject;

/**
 * Factory for creating encryption sessions. Provides a builder API compatible
 * with the canonical godaddy/asherah SessionFactory.
 *
 * <pre>
 * SessionFactory factory = SessionFactory.newBuilder("productId", "serviceId")
 *     .withInMemoryMetastore()
 *     .withNeverExpiredCryptoPolicy()
 *     .withStaticKeyManagementService("masterKey")
 *     .build();
 *
 * try (Session&lt;JSONObject, byte[]&gt; session = factory.getSessionJson("partition")) {
 *     byte[] encrypted = session.encrypt(payload);
 *     JSONObject decrypted = session.decrypt(encrypted);
 * }
 * factory.close();
 * </pre>
 */
public class SessionFactory implements AutoCloseable {

    private final AsherahFactory nativeFactory;

    private SessionFactory(final AsherahFactory nativeFactory) {
        this.nativeFactory = nativeFactory;
    }

    public Session<JSONObject, byte[]> getSessionJson(final String partitionId) {
        return new SessionJsonImpl<>(nativeFactory.getSession(partitionId), false);
    }

    public Session<byte[], byte[]> getSessionBytes(final String partitionId) {
        return new SessionBytesImpl<>(nativeFactory.getSession(partitionId), false);
    }

    public Session<JSONObject, JSONObject> getSessionJsonAsJson(final String partitionId) {
        return new SessionJsonImpl<>(nativeFactory.getSession(partitionId), true);
    }

    public Session<byte[], JSONObject> getSessionBytesAsJson(final String partitionId) {
        return new SessionBytesImpl<>(nativeFactory.getSession(partitionId), true);
    }

    @Override
    public void close() {
        nativeFactory.close();
    }

    public static MetastoreStep newBuilder(final String productId, final String serviceId) {
        return new Builder(productId, serviceId);
    }

    // --- Builder step interfaces (matching canonical API) ---

    public interface MetastoreStep {
        CryptoPolicyStep withInMemoryMetastore();
        CryptoPolicyStep withMetastore(Metastore<JSONObject> metastore);
    }

    public interface CryptoPolicyStep {
        KeyManagementServiceStep withNeverExpiredCryptoPolicy();
        KeyManagementServiceStep withCryptoPolicy(CryptoPolicy cryptoPolicy);
    }

    public interface KeyManagementServiceStep {
        BuildStep withStaticKeyManagementService(String staticMasterKey);
        BuildStep withKeyManagementService(KeyManagementService kms);
    }

    public interface BuildStep {
        BuildStep withMetricsEnabled();
        SessionFactory build();
    }

    // --- Builder implementation ---

    private static final class Builder
            implements MetastoreStep, CryptoPolicyStep, KeyManagementServiceStep, BuildStep {

        private final String productId;
        private final String serviceId;
        private Object metastore;
        private CryptoPolicy cryptoPolicy;
        private KeyManagementService kms;

        Builder(final String productId, final String serviceId) {
            this.productId = productId;
            this.serviceId = serviceId;
        }

        @Override
        public CryptoPolicyStep withInMemoryMetastore() {
            this.metastore = new InMemoryMetastoreImpl<>();
            return this;
        }

        @Override
        public CryptoPolicyStep withMetastore(final Metastore<JSONObject> metastore) {
            if (metastore instanceof InMemoryMetastoreImpl
                    || metastore instanceof JdbcMetastoreImpl
                    || metastore instanceof DynamoDbMetastoreImpl) {
                this.metastore = metastore;
            } else {
                throw new UnsupportedOperationException(
                        "Custom Metastore implementations are not supported by the FFI binding. "
                                + "Use InMemoryMetastoreImpl, JdbcMetastoreImpl, or DynamoDbMetastoreImpl.");
            }
            return this;
        }

        @Override
        public KeyManagementServiceStep withNeverExpiredCryptoPolicy() {
            this.cryptoPolicy = new com.godaddy.asherah.crypto.NeverExpiredCryptoPolicy();
            return this;
        }

        @Override
        public KeyManagementServiceStep withCryptoPolicy(final CryptoPolicy cryptoPolicy) {
            this.cryptoPolicy = cryptoPolicy;
            return this;
        }

        @Override
        public BuildStep withStaticKeyManagementService(final String staticMasterKey) {
            this.kms = new StaticKeyManagementServiceImpl(staticMasterKey);
            return this;
        }

        @Override
        public BuildStep withKeyManagementService(final KeyManagementService kms) {
            this.kms = kms;
            return this;
        }

        @Override
        public BuildStep withMetricsEnabled() {
            // Metrics are handled by the native Rust layer's observability.
            return this;
        }

        @Override
        public SessionFactory build() {
            final AsherahConfig.Builder cb = AsherahConfig.builder()
                    .productId(productId)
                    .serviceName(serviceId);

            // Apply metastore config
            if (metastore instanceof InMemoryMetastoreImpl) {
                ((InMemoryMetastoreImpl<?>) metastore).applyConfig(cb);
            } else if (metastore instanceof JdbcMetastoreImpl) {
                ((JdbcMetastoreImpl) metastore).applyConfig(cb);
            } else if (metastore instanceof DynamoDbMetastoreImpl) {
                ((DynamoDbMetastoreImpl) metastore).applyConfig(cb);
            }

            // Apply crypto policy config
            if (cryptoPolicy != null) {
                cryptoPolicy.applyConfig(cb);
            }

            // Apply KMS config
            if (kms != null) {
                kms.applyConfig(cb);
            }

            final AsherahConfig config = cb.build();
            final AsherahFactory nativeFactory = Asherah.factoryFromConfig(config);
            return new SessionFactory(nativeFactory);
        }
    }
}

File: asherah-dotnet/src/GoDaddy.Asherah.Encryption.Compat/AppEncryption/SessionFactory.cs
using GoDaddy.Asherah.AppEncryption.Crypto;
using GoDaddy.Asherah.AppEncryption.Kms;
using GoDaddy.Asherah.AppEncryption.Persistence;
using GoDaddy.Asherah.Encryption;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption;

/// <summary>
/// Factory for creating encryption sessions. Provides a builder API compatible
/// with the canonical godaddy/asherah SessionFactory.
/// </summary>
public class SessionFactory : IDisposable
{
    private readonly AsherahFactory _nativeFactory;

    private SessionFactory(AsherahFactory nativeFactory)
    {
        _nativeFactory = nativeFactory;
    }

    /// <summary>Returns a JSON-object session with binary UTF-8 encoded data row records.</summary>
    public Session<JObject, byte[]> GetSessionJson(string partitionId)
        => new SessionJsonImpl<byte[]>(_nativeFactory.GetSession(partitionId), false);

    /// <summary>Returns a byte-payload session with binary data row records (JSON ciphertext envelope).</summary>
    public Session<byte[], byte[]> GetSessionBytes(string partitionId)
        => new SessionBytesImpl<byte[]>(_nativeFactory.GetSession(partitionId), false);

    /// <summary>Returns a JSON-object session whose data row records are <see cref="JObject"/>.</summary>
    public Session<JObject, JObject> GetSessionJsonAsJson(string partitionId)
        => new SessionJsonImpl<JObject>(_nativeFactory.GetSession(partitionId), true);

    /// <summary>Returns a byte-payload session whose data row records are <see cref="JObject"/>.</summary>
    public Session<byte[], JObject> GetSessionBytesAsJson(string partitionId)
        => new SessionBytesImpl<JObject>(_nativeFactory.GetSession(partitionId), true);

    /// <summary>Releases the underlying native factory.</summary>
    public void Dispose() => _nativeFactory.Dispose();

    /// <summary>Begins building a <see cref="SessionFactory"/> for the given product and service identifiers.</summary>
    public static IMetastoreStep NewBuilder(string productId, string serviceId)
        => new FactoryBuilder(productId, serviceId);

    // --- Builder step interfaces ---

    /// <summary>Builder step: choose metastore configuration.</summary>
    public interface IMetastoreStep
    {
        /// <summary>Uses the in-process in-memory metastore (testing / dev).</summary>
        ICryptoPolicyStep WithInMemoryMetastore();
        /// <summary>Uses a supported metastore adapter (in-memory, ADO, or DynamoDB).</summary>
        ICryptoPolicyStep WithMetastore(IMetastore<JObject> metastore);
    }

    /// <summary>Builder step: choose crypto policy.</summary>
    public interface ICryptoPolicyStep
    {
        /// <summary>Uses keys that never expire per <see cref="NeverExpiredCryptoPolicy"/>.</summary>
        IKeyManagementServiceStep WithNeverExpiredCryptoPolicy();
        /// <summary>Uses the supplied <paramref name="cryptoPolicy"/>.</summary>
        IKeyManagementServiceStep WithCryptoPolicy(CryptoPolicy cryptoPolicy);
    }

    /// <summary>Builder step: choose key management (KMS).</summary>
    public interface IKeyManagementServiceStep
    {
        /// <summary>Static master key (testing only; sets environment for native static KMS).</summary>
        IBuildStep WithStaticKeyManagementService(string staticMasterKey);
        /// <summary>Uses a supported KMS implementation (static or AWS).</summary>
        IBuildStep WithKeyManagementService(IKeyManagementService kms);
    }

    /// <summary>Final builder step: optional hooks and <see cref="Build"/>.</summary>
    public interface IBuildStep
    {
        /// <summary>Reserved for API parity; metrics are handled by the native layer when configured.</summary>
        IBuildStep WithMetrics(object? metrics);
        /// <summary>Reserved for API parity; logging is handled by the native layer when configured.</summary>
        IBuildStep WithLogger(object? logger);
        /// <summary>Builds the configured factory.</summary>
        SessionFactory Build();
    }

    // --- Builder implementation ---

    private class FactoryBuilder : IMetastoreStep, ICryptoPolicyStep, IKeyManagementServiceStep, IBuildStep
    {
        private readonly string _productId;
        private readonly string _serviceId;
        private object? _metastore;
        private CryptoPolicy? _cryptoPolicy;
        private IKeyManagementService? _kms;

        internal FactoryBuilder(string productId, string serviceId)
        {
            _productId = productId;
            _serviceId = serviceId;
        }

        public ICryptoPolicyStep WithInMemoryMetastore()
        {
            _metastore = new InMemoryMetastoreImpl<JObject>();
            return this;
        }

        public ICryptoPolicyStep WithMetastore(IMetastore<JObject> metastore)
        {
            if (metastore is InMemoryMetastoreImpl<JObject> or AdoMetastoreImpl or DynamoDbMetastoreImpl)
                _metastore = metastore;
            else
                throw new NotSupportedException(
                    "Custom IMetastore implementations are not supported by the FFI binding. " +
                    "Use InMemoryMetastoreImpl, AdoMetastoreImpl, or DynamoDbMetastoreImpl.");
            return this;
        }

        public IKeyManagementServiceStep WithNeverExpiredCryptoPolicy()
        {
            _cryptoPolicy = new NeverExpiredCryptoPolicy();
            return this;
        }

        public IKeyManagementServiceStep WithCryptoPolicy(CryptoPolicy cryptoPolicy)
        {
            _cryptoPolicy = cryptoPolicy;
            return this;
        }

        public IBuildStep WithStaticKeyManagementService(string staticMasterKey)
        {
            _kms = new StaticKeyManagementServiceImpl(staticMasterKey);
            return this;
        }

        public IBuildStep WithKeyManagementService(IKeyManagementService kms)
        {
            _kms = kms;
            return this;
        }

        public IBuildStep WithMetrics(object? metrics) => this; // accepted, handled by native layer
        public IBuildStep WithLogger(object? logger) => this; // accepted, handled by native layer

        public SessionFactory Build()
        {
            var cb = AsherahConfig.CreateBuilder()
                .WithProductId(_productId)
                .WithServiceName(_serviceId);

            // Apply metastore config
            switch (_metastore)
            {
                case InMemoryMetastoreImpl<JObject> mem: mem.ApplyConfig(cb); break;
                case AdoMetastoreImpl ado: ado.ApplyConfig(cb); break;
                case DynamoDbMetastoreImpl dynamo: dynamo.ApplyConfig(cb); break;
            }

            // Apply crypto policy
            _cryptoPolicy?.ApplyConfig(cb);

            // Apply KMS
            _kms?.ApplyConfig(cb);

            var config = cb.Build();
            var nativeFactory = AsherahFactory.FromConfig(config);
            return new SessionFactory(nativeFactory);
        }
    }
}
